x-pack/platform/plugins/shared/alerting/server/backfill_client/backfill_client.ts

```diff
@@ -97,6 +112,107 @@ export class BackfillClient {
     });
   }
 
+  private async deleteAdHocRunsAndTasks({
+    unsecuredSavedObjectsClient,
+    adHocRuns,
+    shouldUpdateGaps,
+    internalSavedObjectsRepository,
+    eventLogClient,
+    eventLogger,
+    actionsClient,
+  }: {
+    unsecuredSavedObjectsClient: SavedObjectsClientContract;
+    adHocRuns: Array<SavedObjectsFindResult<AdHocRunSO>>;
+    shouldUpdateGaps?: boolean;
+    internalSavedObjectsRepository?: ISavedObjectsRepository;
+    eventLogClient?: IEventLogClient;
+    eventLogger?: IEventLogger;
+    actionsClient?: ActionsClient;
+  }) {
+    if (adHocRuns.length === 0) return;
+
+    // Prepare backfill metadata for gap updates before deleting SOs
+    const backfillsForGapUpdate =
+      shouldUpdateGaps && actionsClient && internalSavedObjectsRepository && eventLogClient
+        ? adHocRuns.map((so) =>
+            transformAdHocRunToBackfillResult({
+              adHocRunSO: so,
+              isSystemAction: (id: string) => actionsClient.isSystemAction(id),
+            })
+          )
+        : [];
+
+    const deleteResult = await unsecuredSavedObjectsClient.bulkDelete(
+      adHocRuns.map((adHocRun) => ({
+        id: adHocRun.id,
+        type: AD_HOC_RUN_SAVED_OBJECT_TYPE,
+      }))
+    );
+
+    const deleteErrors = deleteResult.statuses.filter((status) => !!status.error);
+    if (deleteErrors.length > 0) {
+      this.logger.warn(
+        `Error deleting backfill jobs with IDs: ${deleteErrors
+          .map((status) => status.id)
+          .join(', ')} with errors: ${deleteErrors.map(
+          (status) => status.error?.message
+        )} - jobs and associated task were not deleted.`
+      );
+    }
+
+    if (
+      shouldUpdateGaps &&
+      backfillsForGapUpdate.length > 0 &&
+      internalSavedObjectsRepository &&
+      eventLogClient
+    ) {
+      for (const backfill of backfillsForGapUpdate) {
+        if ('rule' in backfill) {
```

Comment:
An early continue will remove this layer of nesting

Reply:
https://github.com/elastic/kibana/pull/244719/commits/502455caad05e70f763bb85fb3f1f220f48cde49